### PR TITLE
created putback option

### PR DESCRIPTION
**Pull request by Christopher-Bradshaw**
Adds a `putback_url` option to the payload. If given, some results are sent as a PUT to there on completion (at the moment that is just the number of diffs). This can be used so a controlling service knows when the science job is done and what results it got.

There is also a bunch of code copied into the new `finalize` package. Sorry for moving things about. Anything that is just moved should be commented as such

### CODE REVIEW

File: finalize/finalize.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"os"
+	"time"
+
+	"github.com/Clever/http-science/config"
+	"github.com/Clever/http-science/science"
+	"github.com/Clever/pathio"
+)
+
+// LogResults logs the results of science to the places given in the config
+func LogResults(startTime time.Time, payload *config.Payload) error {
```
Comment (Christopher-Bradshaw) [inline finalize/finalize.go]: This is just copy pasted out of main except for the stuff re putback at the end

File: finalize/email.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package finalize
```
Comment (Christopher-Bradshaw) [inline finalize/email.go]: Copy pasted out of the email package

File: finalize/error.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package finalize
```
Comment (Christopher-Bradshaw) [inline finalize/error.go]: copy pasted out of config.

File: finalize/finalize.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"os"
+	"time"
+
+	"github.com/Clever/http-science/config"
+	"github.com/Clever/http-science/science"
+	"gopkg.in/Clever/pathio.v3"
+)
+
+// LogResults logs the results of science to the places given in the config
+func LogResults(startTime time.Time, payload *config.Payload) error {
```
Comment (Christopher-Bradshaw) [inline finalize/finalize.go]: Copy pasted out of main except for the last piece about putbacks

**Comment by kshen0** [comment]
LGTM

File: finalize/finalize.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	if payload.PutbackURL != "" && payload.JobType == "correctness" {
+		b, err := json.Marshal(config.PutbackResponse{
+			Diffs: science.Res.Diffs,
+		})
+		if err != nil {
+			return err
+		}
+		req, err := http.NewRequest(http.MethodPut, payload.PutbackURL, bytes.NewReader(b))
```
Comment (kvigen) [inline finalize/finalize.go]: out of curiosity, why put instead of post?

Comment (Christopher-Bradshaw) [inline finalize/finalize.go]: I never really know which of those two to use. . .

**Review by kshen0 — CHANGES_REQUESTED** [review]
I can't find a way in the Github UI to unassign myself from this but it seems likely this PR should be closed?

**Comment by schimmy** [comment]
No @kshen0 this PR must be forever preserved to remember this beautiful moment in time! 😛

**Comment by Christopher-Bradshaw** [comment]
I think it has served its purpose :) I'll close to help @kshen0 get to github inbox 0. Hope you're all doing well!

**Pull request closed without merging**